### ISSUE TITLE
JmDNS returns cached values

### ISSUE THREAD

**Issue by guillerodriguez** [openhab/jmdns#19]
I have a situation where I need to check for services currently available in the network. These services may come and go so I need to recheck, either periodically or when triggered by the user.

However I have found that after JmDNS has discovered a service, it will keep returning cached information regardless of whether the device is still present in the network. I have tried unregistering the listener and registering it again, but when I do this `serviceAdded` is immediately called with stale (cached) information.

Is there any way to force JmDNS to drop any cached information so that I can start a "fresh" search, other than calling `JmDNS.close()`? (this works, but it takes a few seconds to complete).

**Comment by guillerodriguez** [comment]
I can work around this by casting to `JmDNSImpl`, then getting a reference to the cache object and calling `clear()` on it:

    ((JmDNSImpl) jmdns).getCache().clear()

I would suggest to make add a public `getCache()` method in `JmDNS`.

**Comment by spearway** [comment]
This is wrong. The protocol does not work that way. The cache is trimmed by the timeout of the dans entry. If you do that you are not compliant with the standard.

Pierre

**Comment by guillerodriguez** [comment]
If this is wrong, what is the "right" way to force JmDNS to start a new search? Otherwise it just keeps returning stale data (specially since serviceRemoved does not seem to be called when the services disappear, see #18). 

Manually clearing the cache seems to be the only way to make sure the data is up to date.

**Comment by spearway** [comment]
You don’t do this. JmDNS is a DNS service with a cache not a search. This is common misconception on how zero conf works. There is a timeout specified in the std after which the entries are removed and last time I run the conformance test (2 years  ago) it passed without any issues. Killing the cache will cause all sorts of issues on the network as it will force readvertising of the entries. The truce is that with zero con you have to wait for the timeout fto see the device being removed it is the way it works and you should be prepared in your app for the device to fail.

Pierre

**Comment by guillerodriguez** [comment]
"The truce is that with zero con you have to wait for the timeout fto see
the device being removed it is the way it works"

Both Avahi (with avahi-browse or avahi-discover) and Apple's Bonjour (quick
test with the Java bindings in the SDK) immediately notice when a service
is gone, no need to do anything special and no need to wait for any timeout.

But JmDNS does not notice this, or if it does notice, it does not notify
the client application (again, see #18). So no misconceptions here, I am
just trying to find a way to work around this problem (which other
implementations don't have).

El sábado, 14 de febrero de 2015, Pierre Frisch <notifications@github.com>
escribió:

> You don’t do this. JmDNS is a DNS service with a cache not a search. This
> is common misconception on how zero conf works. There is a timeout
> specified in the std after which the entries are removed and last time I
> run the conformance test (2 years ago) it passed without any issues.
> Killing the cache will cause all sorts of issues on the network as it will
> force readvertising of the entries. The truce is that with zero con you
> have to wait for the timeout fto see the device being removed it is the way
> it works and you should be prepared in your app for the device to fail.
>
> Pierre
>
> > On Feb 13, 2015, at 13:36, Guillermo Rodríguez <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
> >
> > If this is wrong, what is the "right" way to force JmDNS to start a new
> search? Otherwise it just keeps returning stale data (specially since
> serviceRemoved does not seem to be called when the services disappear, see
> #18 <https://github.com/openhab/jmdns/issues/18>).
> >
> > Manually clearing the cache seems to be the only way to make sure the
> data is up to date.
> >
> > —
> > Reply to this email directly or view it on GitHub <
> https://github.com/openhab/jmdns/issues/19#issuecomment-74331179>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openhab/jmdns/issues/19#issuecomment-74343180>.
>


-- 
Guillermo Rodriguez Garcia
guille.rodriguez@gmail.com

**Comment by NeoLSN** [comment]
Yeah, I got same problem, too.
And even I call ((JmDNSImpl) jmdns).getCache().clear();.
It still not clear.

**Comment by bigjust12345** [comment]
What appears to be the problem here is jmdns does not allow for setting ttl values when registering services and furthermore appears to ignore ttl value of the records recived instead always forcing records to be valid for an hour

**Comment by namannik** [comment]
I'm seeing this behavior, also.  If a service on the network shuts down gracefully, it broadcasts on the network that it's going away.  On iOS, this correctly causes the service to be removed from the list immediately, but JmDNS seems to hold on to the service.

This is different than the caching mentioned by @spearway .  The cache retains items that were unexpectedly removed from the network.  (i.e. The network cable was unplugged.)

**Comment by tonywestonuk** [comment]
I have also noticed that if I was to, say move a service to a different server.

First, shutting down service on server 1,  JmDNS correctly removes the service.
Starting up same service on server 2.  JmDNS shows new service starting.

However, some of the properties that are retrieved from the new service are cached copies from when it was running on server 1.

I see the change immedately in Bonjour browser,  or zeroconfServiceBrowser (on windows) I see the correct properties.

**Comment by ViToni** [comment]
@guillerodriguez   Could you try out if https://github.com/jmdns/jmdns/pull/126 solves your issue?

**Comment by ViToni** [comment]
Maybe you could try if the release https://github.com/jmdns/jmdns/releases/tag/jmdns-3.5.2 works for you.

If the issue is not reproducible anymore the issue could be closed then.

**Comment by ViToni** [comment]
@guillerodriguez I see. Actually this behaviour is already documented in the spec:
https://tools.ietf.org/html/rfc6762#section-10.4

Unfortunately it is not yet implmented. JmDNS is using currently on expiration (based on TTL of record data).
 
Maybe the title should be changed to reflect the not yet complete implentation of the RFC.

**Comment by ViToni** [comment]
Most record data have an TTL of 120 so will be removed latest after that time.
The most prominent exception is the PTR record whcih has a TTL of 4500s / 75 min which might be part of the behaviour you see.

**Comment by guillerodriguez** [comment]
> Maybe the title should be changed to reflect the not yet complete implentation of the RFC.

@ViToni That's right. Just changed the title.

**Comment by britishd** [comment]
@ViToni 

Hey. Please look into this issue of `cordova` plugin that use Jmnds implementation. 

[Cached data still living after unregister device](https://github.com/becvert/cordova-plugin-zeroconf/issues/38)

We have same problem with duplicate data. Plugin use the latest release of Jmdns. May be you can help me with this problem.

**Comment by nmpribeiro** [comment]
Also noticed this behaviour. Trying to get my service to be unregistered right after (or at least a couple seconds after) my cable is unplugged.
Closing and re-opening jmdns is really awkward. I would prefer to ask the user for a 'refresh' button...

**Comment by farble1670** [comment]
Sorry for the necro, but using 3.5.7 has the same issue, and calling `JmDNSImpl.clearCache()` does not resolve the issue. Not that it probably matters but this is on Android. I don't see a solution other than periodically closing and recreating the `JmDNS` instance.

The stale data definitely persists greater than 120 seconds.

Note I'm using `JmDNS.list()`, not `JmDNS.addServiceListener()`.
